Mathlib/Data/Polynomial/Derivation.lean

```diff
@@ -92,4 +92,62 @@ def mkDerivationEquiv : A ≃ₗ[R] Derivation R R[X] A :=
 @[simp] lemma mkDerivationEquiv_symm_apply (D : Derivation R R[X] A) :
     (mkDerivationEquiv R).symm D = D X := rfl
 
+end CommSemiring
 end Polynomial
+
+namespace Derivation
+
+variable {R A M : Type*} [CommSemiring R] [CommSemiring A] [Algebra R A] [AddCommMonoid M]
+  [Module A M] [Module R M] [IsScalarTower R A M] (d : Derivation R A M) (a : A)
+
+open Polynomial Module
+
+/--
+For a derivation `d : A → M` and an element `a : A`, `d.comp_aeval a` is the
+derivation of `R[X]` which takes a polynomial `f` to `d(aeval a f)`.
+
+This derivation takes values in `Module.AEval R M a`, which is `M`, regarded as an
+`R[X]`-module, with the action of a polynomial `f` defined by `f • m = (aeval a f) • m`.
+-/
+/-
+Note: It is currently not easy to define `comp_aeval` using `Derivation.compAlgebraMap`.
+This is partly because `A` is not an `R[X]` algebra and partly because `Derivation.compAlgebraMap`
+is only implemented in the case that `R` is a `CommRing`, not a `CommSemiring`.
+-/
+def comp_aeval : Derivation R R[X] <| AEval R M a where
+  toFun f          := AEval.of R M a (d (aeval a f))
+  map_add'         := by simp
+  map_smul'        := by simp
+  leibniz'         := by simp [AEval.of_aeval_smul]
+  map_one_eq_zero' := by simp
+
+lemma comp_aeval_def (d : Derivation R A M) (f : R[X]) :
+    d.comp_aeval a f = AEval.of R M a (d (aeval a f)) := rfl
+
+lemma comp_aeval_apply (d : Derivation R A M) (f : R[X]) :
```

Comment:
I would argue that this lemma is most likely only going to be used once.
In all other cases, it will probably be used in the reverse direction, and `simp` will take care of that.
So I think you can just inline this into the proof of `comp_aeval_eq`, possibly using a `calc`-block.

Reply:
Thanks you for the suggestion.